### ISSUE TITLE
Decide which prototype parser to use

### ISSUE THREAD

**Issue by jpschorr** [partiql/partiql-lang-rust#76]
Currently, there are two prototype parsers: a PEG parser (using Pest) and an LR parser (generated by LALRPop).

We need to 
 - evaluate which parser to continue development with
 - remove the other parser
 - clean up the exported API

**Comment by jpschorr** [comment]
Output of `cargo bench`

```bash
     Running unittests (/Volumes/workplace/partiql-lang-rust/target/release/deps/bench_parse-149da44531c5e34d)
WARNING: HTML report generation will become a non-default optional feature in Criterion.rs 0.4.0.
This feature is being moved to cargo-criterion (https://github.com/bheisler/cargo-criterion) and will be optional in a future version of Criterion.rs. To silence this warning, either switch to cargo-criterion or enable the 'html_reports' feature in your Cargo.toml.

peg-simple              time:   [2.5382 us 2.6437 us 2.7640 us]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

peg-ion                 time:   [32.513 us 36.987 us 43.082 us]
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe

peg-group               time:   [47.252 us 52.460 us 58.732 us]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

peg-complex             time:   [1.0547 ms 1.0717 ms 1.0924 ms]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

peg-ast-simple          time:   [3.0249 us 3.0464 us 3.0717 us]
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe

peg-ast-ion             time:   [13.548 us 13.596 us 13.656 us]
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  6 (6.00%) high severe

peg-ast-group           time:   [36.328 us 36.924 us 37.599 us]
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe

peg-ast-complex         time:   [1.0424 ms 1.0472 ms 1.0536 ms]
Found 15 outliers among 100 measurements (15.00%)
  5 (5.00%) high mild
  10 (10.00%) high severe

logos-simple            time:   [267.03 ns 268.88 ns 271.08 ns]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe

logos-ion               time:   [564.48 ns 567.41 ns 570.89 ns]
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  8 (8.00%) high severe

logos-group             time:   [2.3590 us 2.3804 us 2.4057 us]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

logos-complex           time:   [4.8537 us 4.8805 us 4.9137 us]
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe

lalr-simple             time:   [556.06 ns 558.06 ns 560.71 ns]
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  5 (5.00%) high mild
  2 (2.00%) high severe

lalr-ion                time:   [2.6945 us 2.7222 us 2.7611 us]
Found 15 outliers among 100 measurements (15.00%)
  6 (6.00%) high mild
  9 (9.00%) high severe

lalr-group              time:   [9.3843 us 9.4308 us 9.4879 us]
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) high mild
  7 (7.00%) high severe

lalr-complex            time:   [20.159 us 20.201 us 20.250 us]
Found 16 outliers among 100 measurements (16.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  14 (14.00%) high severe
```

**Comment by jpschorr** [comment]
For me, the takeaway from the above is that you can see the penalty we pay for the PEG's backtracking as the query complexity increases. It's possible that we could rework the PEG rules to alleviate some of this, but SQL and PartiQL syntax does not have the frequent commas, semicolons, brackets, and braces of C-style languages that often alleviate this.

```
peg-ast-simple          time:   [3.0249 us 3.0464 us 3.0717 us]
peg-ast-complex         time:   [1.0424 ms 1.0472 ms 1.0536 ms]

lalr-simple             time:   [556.06 ns 558.06 ns 560.71 ns]
lalr-complex            time:   [20.159 us 20.201 us 20.250 us]
```

**Comment by jpschorr** [comment]
It should also be noted that the PEG parser is not quite correct even for the tests it passes. 

Logical and mathematical expressions require parsing with implicit whitespace. E.g., `3+4>7` should be parsed the same as `      3+     4 > 7`.

But SQL/PartiQL's text keywords require whitespace to separate. E.g., `SELECT a` should parse, but `SELECTa` should be a syntax error.

Currently, the PEG parser has implicit whitespace everywhere, even between text. E.g., currently `SELECT*FROMaLIMIT10` parses even though it should not.

**Comment by alancai98** [comment]
I agree with what @jpschorr had mentioned in his comments. The performance and implicit whitespace parsing behaviors are some notable drawbacks from the PEG parser. Also, the PEG (at least using [pest.rs](https://pest.rs/)) requires the additional pass to convert to the AST (see [ast_builder.rs](https://github.com/partiql/partiql-lang-rust/blob/3380e456cdf04ce11d134a2a5416b30fef5c83ab/partiql-parser/src/peg/ast_builder.rs)), further adding to the parse time and requiring additional conversion code for new syntax.

For this discussion (and maybe for docs we export to `partiql-docs` for instance), we may want to also identify some cons from deciding to use the LALR parser. I'm still learning more about parsing approaches, but from our prior discussions, LALR
- error reporting may be trickier
- harder to embed an LR grammar into another LR grammar
  - embedding grammars may not be too relevant for PartiQL's use cases
- learning curve without prior experience (e.g. knowing when to shift-reduce)

To me, the performance benefits along with the other pros outweigh the cons of using the LALR parser.

---

As a sidenote, I believe @am357 had done some prior research and found SQL parsers have generally used LALR (e.g. [SQLite](https://sqlite.org/lemon.html) uses a LALR(1) parser generator called Lemon).

**Comment by am357** [comment]
This note follows what @jpschorr and @alancai98 have already mentioned in previous comments:

I also agree with the drawbacks mentioned especially the performance one. In addition, I think another aspect that we need to look at is the amount of changes that we need to make if we need to re-adjust the grammar in early stages or introduce a new syntax later on (if required); from what I've seen so far, between two implementations, `LALRPop` appears to require less work because we don't need to modify an additional pass.

Having said that, ideally, I see value in coming up with an overall score for each parser based on a set of relevant criteria (with their corresponding weights). Maybe we can have a quick brainstorming session for this?

@alancai98 by the way, to be honest, my research hasn't been exhaustive. Apart from what you've already mentioned, I've also come across [this chapter](https://www.oreilly.com/library/view/flex-bison/9780596805418/ch04.html) that is using BSON (supposedly with its default LALR(1)) parser (See _SQL to RPN_ section). In contrast [sqlparser-rs](https://github.com/sqlparser-rs/sqlparser-rs) uses a TDOP parser for its expression parsing while using a recursive descent parser for statements; seemingly both handcrafted (See _Design_ in README).

**Comment by jpschorr** [comment]
Aside from the tradeoffs mentioned above, there is also compile time.

The PEG parser compiles in seconds or less.

The LALRPOP `build.rs` step takes ~152 seconds for a clean build to generate the parser rust code from the `lalrpop` file. Once generated, subsequent compiles are fast as long as the parser doesn't need to be regenerated. We could mitigate this by changing the LALRPOP generation step to an explicit step and check in the generated file, as the LALRPOP project itself does:
- https://github.com/lalrpop/lalrpop/tree/master/lalrpop/src/parser
- https://github.com/lalrpop/lalrpop/blob/master/update_lrgrammar.sh

Once work on the parser/grammar settles down, this mitigation will likely be a good thing to do anyway; otherwise dependencies pulling in partiql crates will have to pay the generation cost.
